eo-maven-plugin/src/main/java/org/eolang/maven/VersionsMojo.java

```diff
@@ -0,0 +1,141 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.eolang.maven;
+
+import com.jcabi.log.Logger;
+import com.jcabi.xml.XMLDocument;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.cactoos.Text;
+import org.cactoos.experimental.Threads;
+import org.cactoos.iterable.Filtered;
+import org.cactoos.iterable.IterableOf;
+import org.cactoos.iterable.Mapped;
+import org.cactoos.list.Joined;
+import org.cactoos.list.ListOf;
+import org.cactoos.number.SumOf;
+import org.cactoos.scalar.Reduced;
+import org.cactoos.set.SetOf;
+import org.cactoos.text.Replaced;
+import org.cactoos.text.TextOf;
+import org.cactoos.text.UncheckedText;
+import org.eolang.maven.hash.CommitHash;
+import org.eolang.maven.hash.CommitHashesMap;
+import org.eolang.maven.tojos.ForeignTojo;
+import org.eolang.maven.util.Home;
+
+/**
+ * Mojo that replaces tags with hashes in XMIR.
+ *
+ * @since 0.29.6
+ * @todo #1602:30min Handle tags that are not in available versions list.
+ *  VersionsMojo goes right after OptimizeMojo and replaces all tags with
+ *  comparable hashes. EO code may contains tags that are not in available
+ *  versions list (see: <a href="https://home.objectionary.com/tags.txt"/>).
+ *  We need to catch somehow such versions and throw an exception. Or we can
+ *  place the VersionsMojo right after ParseMojo and create new xsl which is
+ *  used on optimization step and caches such invalid tags.
+ */
+public final class VersionsMojo extends SafeMojo {
+    /**
+     * Tag pattern.
+     */
+    private static final Pattern SEMVER = Pattern.compile("[0-9]+\\.[0-9]+\\.[0-9]+");
+
+    /**
+     * Commit hashes map.
+     */
+    @Parameter(required = true, property = "eo.commitHashes")
+    private final Map<String, CommitHash> hashes = new CommitHashesMap();
+
+    @Override
+    void exec() throws IOException {
+        if (this.withVersions) {
+            final Collection<ForeignTojo> tojos = this.scopedTojos().notDiscovered();
+            final Path dir = this.targetDir.toPath();
+            final String format = "ver=\"%s\"";
+            final int total = new SumOf(
+                new Threads<>(
+                    Runtime.getRuntime().availableProcessors(),
+                    new Mapped<>(
+                        tojo -> () -> {
+                            final Path path = tojo.optimized();
+                            final Text source = new UncheckedText(new TextOf(path));
+                            final List<Text> tags = new ListOf<>(
```

Comment:
@maxonfjvipon It's just a question. The `tags` variable could be actually inlined, but I'm not sure how the code will look after it (will it be more readable?). What do you think?

Reply:
@maxonfjvipon I suggest to leave it as is :)